### PR TITLE
Player and ExternalData service structure

### PR DESCRIPTION
**Pull request by Vinnehboom**

### CODE REVIEW

**Comment by Vinnehboom** [comment]
I've decided to keep the `cost` attribute off of the `Player` for now; I think it will be best for historic representation to make a separate model out of this with a `start` and `end`, with the current `Score` being the one without and `end`. 

When looking at historic draft results, we can easily show the score that was active for a tournament or a draft's date

File: app/services/external_data/pokemon/tcg/players.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+module ExternalData
+
+  module Pokemon
+
+    module Tcg
+
+      class Players
+
+        include HTTParty
+        base_uri 'https://limitlesstcg.com'
```
Comment (NemoRoberAmotek) [inline app/services/external_data/pokemon/tcg/players.rb:10]: I'd put this in an environment variable/config of some sorts

File: app/services/external_data/pokemon/tcg/players.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        def self.all
+          get_from(players: [], page: 1)
+        end
+
+        def self.get_players_by(region: 'all', time: 'all', game_format: 'all', show: 100, page: 1)
```
Comment (NemoRoberAmotek) [inline app/services/external_data/pokemon/tcg/players.rb:16]: This feels like an unnecessary method

File: app/services/external_data/pokemon/tcg/players.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+          get('/players', query: { region:, time:, format: game_format, type: 'all', show:, page: })
+        end
+
+        def self.get_from(players: [], page: 1)
```
Comment (NemoRoberAmotek) [inline app/services/external_data/pokemon/tcg/players.rb:20]: i'd rename this method, it sounds to me like it fetches all records from a certain page, while it actually fetches all records from that page and up, if I read it correctly

**Review by NemoRoberAmotek — APPROVED** [review]

File: app/models/game.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class Game < ApplicationRecord
+
+  validates :name, presence: true
+  validates :base_uri, presence: true
```
[the remark this answers is not in the document]
Comment (Vinnehboom) [reply app/models/game.rb:4]: The `Game` is as you guessed below: it's the "Esport/ TCG" that we're covering. The base uri in this case is where we find the data, be it players, tournaments,...

File: app/exceptions/custom_exception.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+class CustomException < StandardError
```
[the remark this answers is not in the document]
Comment (Vinnehboom) [reply app/exceptions/custom_exception.rb:1]: It's true, my reasoning was for these errors to be more identifiable when popping up in our server logs or error reporting and we'd immediately know it's an exception we raised manually.

File: app/models/game.rb
[the remark this answers is not in the document]
Comment (Vinnehboom) [reply app/models/game.rb:1]: Only one record will exist indeed. It was a bit of a question still for me how to handle this; it does come with some hardcoded aspects. I considered making it a non-db model, but I felt there was too much logic we'd eventually want to link to this later down the road so it has become a proper model for now

File: app/services/external_data/pokemon/tcg/players.rb (hunk 3 above)
Comment (Vinnehboom) [reply app/services/external_data/pokemon/tcg/players.rb:20]: Yeah, naming here wasn't convenient. In essence this is the recursive method we need to run the parser. I'll give it another shot

File: app/exceptions/custom_exception.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (NemoRoberAmotek) [reply app/exceptions/custom_exception.rb:1]: Mhm right, in that case i'd rename it to something project-specific `Tcg::Exception` or something

**Pull request merged**